File: Aula_T05/MinhaData.py
class MinhaData:
    
    def __init__(self, dia=1, mes=1, ano=1970, outraData=None, interval=0):
        #if outraData:
        if outraData and isinstance(outraData, MinhaData):
            if interval > 0:
                pass
            else:
                self.__dia = outraData.__dia
                self.__mes = outraData.mes
                self.__ano = outraData.ano
        elif MinhaData.dataValida(dia, mes, ano):
            self.__dia = dia
            self.__mes = mes
            self.__ano = ano

    def inicializaData(self, dia=1, mes=1, ano=1970, outraData=None, interval=0):
        
        if outraData and isinstance(outraData, MinhaData):
            if interval > 0:
                pass  
            else:
                self.__dia = outraData.dia
                self.__mes = outraData.mes
                self.__ano = outraData.ano
                return True
        elif MinhaData.dataValida(dia, mes, ano):
            self.__dia = dia
            self.__mes = mes
            self.__ano = ano
            return True
        return False

    # Implementando set e get dos atributos
    @property
    def dia(self):
        return self.__dia
    
    @dia.setter
    def dia(self, dia):
        if MinhaData.dataValida(dia, self.__mes, self.__ano):
            self.__dia = dia

    @property
    def mes(self):
        return self.__mes
    
    @mes.setter
    def mes(self, mes):
        if MinhaData.dataValida(self.__dia, mes, self.__ano):
            self.__mes = mes

    @property
    def ano(self):
        return self.__ano        
    
    @ano.setter
    def ano(self, ano):
        if MinhaData.dataValida(self.__dia, self.__mes, ano):
            self.__ano = ano
    
    def dataValida(dia, mes, ano):
        if ano < 1900 or ano > 2050:
            return False
        if dia < 1 or dia > 31:
            return False
        if mes == 2:
            if ano % 4 == 0 and (ano % 100 != 0 or ano % 400 == 0):
                if dia <= 29:
                    return True
                else:
                    return False
            else:
                if dia <= 28:
                    return True
                else:
                    return False
        elif mes == 4 or mes == 6 or mes == 9 or mes == 11:
            if dia <= 30:
                return True
            else:
                return False
        elif mes == 1 or mes == 3 or mes == 5 or mes == 7 or mes == 8 or mes == 10 or mes == 12:
            if dia <= 31:
                return True
            else:
                return False
        else:
            return False
        
    def __dataValida(self):
        return MinhaData.dataValida(self.__dia, self.__mes, self.__ano)
        
    def __str__(self):
        return f'{self.dia}/{self.mes}/{self.ano}'

    def __eq__(self, other):
        return self.dia == other.dia and self.mes == other.mes and self.ano == other.ano

    def __ne__(self, other):
        return self.dia != other.dia or self.mes != other.mes or self.ano != other.ano

    def __lt__(self, other):
        if self.ano < other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes < other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia < other.dia:
                    return True
        return False

    def __le__(self, other):
        if self.ano < other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes < other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia < other.dia or self.dia == other.dia:
                    return True
        return False

    def __gt__(self, other):
        if self.ano > other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes > other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia > other.dia:
                    return True
        return False

    def __ge__(self, other):
        if self.ano > other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes > other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia > other.dia or self.dia == other.dia:
                    return True
        return False
    

class DataSemestre:
    
    def __init__(self, dia=1, mes=1, ano=1970, outraData=None, interval=0, desc = 'A Origem de Tudo'):
            #if outraData:
        if outraData and isinstance(outraData, MinhaData):
            if interval > 0:
                pass
            else:
                self.__dia = outraData.__dia
                self.__mes = outraData.mes
                self.__ano = outraData.ano
        elif MinhaData.dataValida(dia, mes, ano):
            self.__dia = dia
            self.__mes = mes
            self.__ano = ano

        self.__desc = desc

    def inicializaData(self, dia=1, mes=1, ano=1970, outraData=None, interval=0):
        
        if outraData and isinstance(outraData, MinhaData):
            if interval > 0:
                pass  
            else:
                self.__dia = outraData.dia
                self.__mes = outraData.mes
                self.__ano = outraData.ano
                return True
        elif MinhaData.dataValida(dia, mes, ano):
            self.__dia = dia
            self.__mes = mes
            self.__ano = ano
            return True
        return False

    # Implementando set e get dos atributos
    @property
    def dia(self):
        return self.__dia
    
    @dia.setter
    def dia(self, dia):
        if MinhaData.dataValida(dia, self.__mes, self.__ano):
            self.__dia = dia

    @property
    def mes(self):
        return self.__mes
    
    @mes.setter
    def mes(self, mes):
        if MinhaData.dataValida(self.__dia, mes, self.__ano):
            self.__mes = mes

    @property
    def ano(self):
        return self.__ano        
    
    @ano.setter
    def ano(self, ano):
        if MinhaData.dataValida(self.__dia, self.__mes, ano):
            self.__ano = ano

    @property
    def desc(self):
        return self.__desc
    
    @desc.setter
    def desc(self, desc):
        self.__desc = desc
    
    def dataValida(dia, mes, ano):
        if ano < 1900 or ano > 2050:
            return False
        if dia < 1 or dia > 31:
            return False
        if mes == 2:
            if ano % 4 == 0 and (ano % 100 != 0 or ano % 400 == 0):
                if dia <= 29:
                    return True
                else:
                    return False
            else:
                if dia <= 28:
                    return True
                else:
                    return False
        elif mes == 4 or mes == 6 or mes == 9 or mes == 11:
            if dia <= 30:
                return True
            else:
                return False
        elif mes == 1 or mes == 3 or mes == 5 or mes == 7 or mes == 8 or mes == 10 or mes == 12:
            if dia <= 31:
                return True
            else:
                return False
        else:
            return False
        
    def __dataValida(self):
        return MinhaData.dataValida(self.__dia, self.__mes, self.__ano)
        
    def __str__(self):
        return f'{self.dia}/{self.mes}/{self.ano}: {self.desc}'

    def __eq__(self, other):
        return self.dia == other.dia and self.mes == other.mes and self.ano == other.ano

    def __ne__(self, other):
        return self.dia != other.dia or self.mes != other.mes or self.ano != other.ano

    def __lt__(self, other):
        if self.ano < other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes < other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia < other.dia:
                    return True
        return False

    def __le__(self, other):
        if self.ano < other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes < other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia < other.dia or self.dia == other.dia:
                    return True
        return False

    def __gt__(self, other):
        if self.ano > other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes > other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia > other.dia:
                    return True
        return False

    def __ge__(self, other):
        if self.ano > other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes > other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia > other.dia or self.dia == other.dia:
                    return True
        return False


class DataSemestreC:

    def __init__(self, dia=1, mes=1, ano=1970, outraData=None, interval=0, desc = 'A Origem de Tudo'):
        self.__data = MinhaData(dia, mes, ano, outraData, interval)
        self.__desc = desc

    def inicializaData(self, dia=1, mes=1, ano=1970, outraData=None, interval=0):
        return self.__data.inicializaData(dia, mes, ano, outraData, interval)

    # Implementando set e get dos atributos
    @property
    def dia(self):
        return self.__data.dia
    
    @dia.setter
    def dia(self, dia):
        self.__data.dia = dia

    @property
    def mes(self):
        return self.__data.mes
    
    @mes.setter
    def mes(self, mes):
        self.__data.mes = mes

    @property
    def ano(self):
        return self.__data.ano        
    
    @ano.setter
    def ano(self, ano):
        self.__data.ano = ano

    @property
    def desc(self):
        return self.__desc
    
    @desc.setter
    def desc(self, desc):
        self.__desc = desc
    
    def dataValida(dia, mes, ano):
        if ano < 1900 or ano > 2050:
            return False
        if dia < 1 or dia > 31:
            return False
        if mes == 2:
            if ano % 4 == 0 and (ano % 100 != 0 or ano % 400 == 0):
                if dia <= 29:
                    return True
                else:
                    return False
            else:
                if dia <= 28:
                    return True
                else:
                    return False
        elif mes == 4 or mes == 6 or mes == 9 or mes == 11:
            if dia <= 30:
                return True
            else:
                return False
        elif mes == 1 or mes == 3 or mes == 5 or mes == 7 or mes == 8 or mes == 10 or mes == 12:
            if dia <= 31:
                return True
            else:
                return False
        else:
            return False
        
    def __dataValida(self):
        return MinhaData.dataValida(self.__dia, self.__mes, self.__ano)
        
    def __str__(self):
        return f'{self.__data}: {self.desc}'

    def __eq__(self, other):
        return self.dia == other.dia and self.mes == other.mes and self.ano == other.ano

    def __ne__(self, other):
        return self.dia != other.dia or self.mes != other.mes or self.ano != other.ano

    def __lt__(self, other):
        if self.ano < other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes < other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia < other.dia:
                    return True
        return False

    def __le__(self, other):
        if self.ano < other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes < other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia < other.dia or self.dia == other.dia:
                    return True
        return False

    def __gt__(self, other):
        if self.ano > other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes > other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia > other.dia:
                    return True
        return False

    def __ge__(self, other):
        if self.ano > other.ano:
            return True
        elif self.ano == other.ano:
            if self.mes > other.mes:
                return True
            elif self.mes == other.mes:
                if self.dia > other.dia or self.dia == other.dia:
                    return True
        return False
    

class DataSemestreE(MinhaData):

    def __init__(self, dia=1, mes=1, ano=1970, outraData=None, interval=0, desc = 'A Origem de Tudo'):
        super().__init__(dia, mes, ano, outraData, interval)
        self.__desc = desc


    @property
    def desc(self):
        return self.__desc
    
    @desc.setter
    def desc(self, desc):
        self.__desc = desc
    
    
    def __str__(self):
        return f'{super().__str__()}: {self.desc}'
    
    

File: Aula_P02/DataPack/MinhaData.java
package DataPack;

public class MinhaData {
    /*
     * A quantidade de dias de cada mês, considerando meses com 30 ou 31 dias e fevereiro com 29 ou 29 dias.
     * Meses entre janeiro (mês 1) e dezembro (mês 12)
     * Anos desde 1900 até 2050
     */
    private byte dia;
    private byte mes;
    private short ano;

    // Cria uma instância com a data como 1/1/1970.
    public MinhaData(){
        this.inicializaData();
    }
    /*
     * Se estes parâmetros não caracterizarem uma data válidos 
     * atributos da classe são inicializados com 1/1/1970;
     */
    public MinhaData(byte dia, byte mes, short ano)
    {
        this();
        inicializaData(dia, mes, ano);
    }

    /*
     * Inicializa uma data com os mesmos atributos da instância da 
     * classe MinhaData. 
     */
    public MinhaData(MinhaData outraData)
    {
        this.inicializaData(outraData);
    }

    /*
     * Inicializa uma data com a data definida na instância da classe 
     * MinhaData incrementada na quantidade de dias definidas pelo 
     * parâmetro intervalo.
     */
    public MinhaData(MinhaData outraData, int interval)
    {
        this(outraData);
        inicializaData(outraData, interval);

    }

    /*
     * Inicializa uma data com a data definida na instância da 
     * classe MinhaData incrementada na quantidade de dias 
     * definidas pelo parâmetro intervalo.
     */
    public boolean inicializaData(MinhaData outraData, int dias)
    {
        byte dia = outraData.dia;
        byte mes = outraData.mes;
        short ano = outraData.ano;
        for (int i = 0; i < dias; i++) {
            dia++;
            if (!dataValida(dia, mes, ano)) {
                dia = 1;
                mes++;
                if (!dataValida(dia, mes, ano)) {
                    mes = 1;
                    ano++;
                    if(!dataValida(dia, mes, ano)) {
                        return false;
                    }
                }        
            }
        }
        this.inicializaData(dia, mes, ano);
        return true;
    }

    /*
     * Inicializa uma data com os mesmos atributos da instância da 
     * classe MinhaData.
     */
    public void inicializaData(MinhaData outraData)
    { 
        this.dia = outraData.dia;
        this.mes = outraData.mes;
        this.ano = outraData.ano;
    }

    /*
     * Inicializa uma data com os valores definidos pelos parâmetros de 
     * entrada. Se estes parâmetros não caracterizarem uma data válidos, 
     * os atributos da classe não são alterados e o método retorna false; 
     */
    public boolean inicializaData(byte dia, byte mes, short ano)
    {
        if (!dataValida(dia, mes, ano)) {
            return false;
        }
        this.dia = dia;
        this.mes = mes;
        this.ano = ano;
        return true;
        
    }
    
    public static boolean dataValida(byte d, byte m, short a) 
    {
        byte diasMes[] = {31,28,31,30,31,30,31,31,30,31,30,31};

        if (a < 1900 || a > 2050) {
            return false;
        }

        if (m < 1 || m > 12) {
            return false;
        }

        if (d < 1 || d > diasMes[m-1]) {
            return false;
        }

        // ano bissexto
        if (m == 2 && d == 29) {
            if (a % 4 == 0 && (a % 100 != 0 || a % 400 == 0)) {
                return true;
            } else {
                return false;
            }
        }
        return true;
    }

    private boolean dataValida()
    {
        return dataValida(this.dia, this.mes, this.ano);
    }

    // Inicializa a data como 1/1/1970;
    public void inicializaData()
    {
        this.dia = 1;
        this.mes = 1;
        this.ano = 1970;
    }

    public String toString(){
        return this.dia + "/" + this.mes + "/" + this.ano;
    }

    public byte qualDia(){
        return this.dia;
    }

    public byte qualMes(){
        return this.mes;
    }

    public short qualAno(){
        return this.ano;
    }

    public boolean setDia(byte dia){
        if (dataValida(dia, this.mes, this.ano)) {
            this.dia = dia;
            return true;
        }
        return false;
    }

    public boolean setMes(byte mes){
        if (dataValida(this.dia, mes, this.ano)) {
            this.mes = mes;
            return true;
        }
        return false;
    }

    public boolean setAno(short ano){
        if (dataValida(this.dia, this.mes, ano)) {
            this.ano = ano;
            return true;
        }
        return false;
    }

    public boolean igualA(MinhaData outraData)
    {
        return this.dia == outraData.dia && this.mes == outraData.mes && this.ano == outraData.ano;

    }
    
    public boolean diferenteDe(MinhaData outraData)
    {
        return !igualA(outraData);

    }

    public boolean anteriorA(MinhaData outraData)
    {
        if (this.ano < outraData.ano) {
            return true;
        } else if (this.ano == outraData.ano) {
            if (this.mes < outraData.mes) {
                return true;
            } else if (this.mes == outraData.mes) {
                if (this.dia < outraData.dia) {
                    return true;
                }
            }
        }
        return false;

    }

    public boolean posteriorA(MinhaData outraData)
    {
        return !anteriorA(outraData) && !igualA(outraData);

    }


}